PROTOCOL.md

```diff
@@ -0,0 +1,563 @@
+
+For example, the [Delta Sink for Apache Spark's Structured Streaming](https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSink.scala) ensures exactly-once semantics when writing a stream into a table using the following process:
+ 1. Record in a write-ahead-log the data that will be written, along with a monotonically increasing identifier for this batch.
+ 2. Check the current version of the transaction with `appId = streamId` in the target table. If this value is greater than or equal to the batch being written, then this data has already been added to the table and processing can skip to the next batch.
+ 3. Write the data optimistically into the table.
+ 4. Attempt to commit the transaction containing both the addition of the data written out and an updated `appId` `version` pair.
+
+The semantics of the application-specific `version` are left up to the external system.
+Delta only ensures that the latest `version` for a given `appId` is available in the table snapshot.
+The Delta transaction protocol does not, for example, assume monotonicity of the `version` and it would be valid for the `version` to decrease, possibly representing a "rollback" of an earlier transaction.
+
+The schema of the `txn` action is as follows:
+
+Field Name | Data Type | Description
+-|-|-
+appId | String | A unique identifier for the application performing the transaction
+version | Long | An application-specific numeric identifier for this transaction
+
+The following is an example `txn` action:
+```
+{
+  "txn": {
+    "appId":"3ba13872-2d47-4e17-86a0-21afd2a22395",
+    "version":364475
+  }
+}
+```
+
+### Protocol Evolution
+The `protocol` action is used to increase the version of the Delta protocol that is required to read or write a given table.
+Protocol versioning allows a newer client to exclude older readers and/or writers that are missing features required to correctly interpret the transaction log.
+The _protocol version_ will be increased whenever non-forward-compatible changes are made to this specification.
+In the case where a client is running an invalid protocol version, an error should be thrown instructing the user to upgrade to a newer protocol version of their Delta client library.
+
+Since breaking changes must be accompanied by an increase in the protocol version recorded in a table, clients can assume that unrecognized fields or actions are never required in order to correctly interpret the transaction log.
+
+The schema of the `protocol` action is as follows:
+
+Field Name | Data Type | Description
+-|-|-
+minReaderVersion | Int | The minimum version of the Delta read protocol that a client must implement in order to correctly *read* this table
+minWriterVersion | Int | The minimum version of the Delta write protocol that a client must implement in order to correctly *write* this table
```

Comment:
Is it worth pointing to where in the code the latest version is specified? Do we have a plan to show a history of protocol upgrades when we add new ones?